packages/compose/src/hooks/use-drop-zone/index.js

```diff
@@ -61,11 +63,16 @@ export default function useDropZone( {
 	const onDragOverRef = useFreshRef( _onDragOver );
 
 	return useRefEffect(
-		( element ) => {
+		( elem ) => {
 			if ( isDisabled ) {
 				return;
 			}
 
+			// If a custom dropZoneRef is passed, use that instead of the element.
+			// This allows the dropzone to cover an expanded area, rather than
+			// be restricted to the area of the ref returned by this hook.
+			const element = dropZoneRef?.current ?? elem;
```

Comment:
I'm wondering how this will work in terms of the cleanup of the hook.

Because it's a ref effect, I think the callback will typically only trigger when `elem` is unmounted, not when `dropZoneRef.current` is unmounted.

Reply:
Good question! I think this is working pretty well so far. To confirm I logged out `element` within the callback, and it appears to be the right element:

<img width="445" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/dede0dbd-0816-4005-b76d-3afba45e756e">

As far as I can tell the behaviour is something like the following:

* We pass in a ref to `useDropZone`, but `useDropZone` also returns a `ref`. That `ref` is still used in the list view, to attach to the `TreeGrid` component.
* When `TreeGrid` unmounts, that results in the cleanup callback being called.
* Because the cleanup callback was instantiated with a reference to `element` being the one passed in by `dropZoneRef` the cleanup callback cleans up the passed in element.
* Also, because we have `dropZoneRef?.current` in the dependencies array of `useRefEffect` when the passed in ref changes, the cleanup appears to correctly fire, too. From checking the docs, this appears to be consistent with [how useEffect works](https://react.dev/reference/react/useEffect#my-cleanup-logic-runs-even-though-my-component-didnt-unmount) in that the cleanup fires not only when the component unmounts, but whenever any of the dependencies change.

So, all up, I don't think there are any stray event listeners not being removed, as far as I can tell.

Do let me know if you think I've missed anything here, though!

Reply:
> When TreeGrid unmounts, that results in the cleanup callback being called.

Sorry if I'm making this more complicated. Given this is being introduced as a generic stable API, I'm mainly thinking that there could be different ways it's used (or misused):
- The `dropZoneRef` could reference a completely separate part of the DOM that isn't mounted/unmounted at the same time as the effect.
- The `dropZoneRef` could reference a parent of the effect that isn't mounted/unmounted at the same time.

I don't know if there are tests that can be added to assert these different cases work as expected.

It also might be worth expanding upon the docs for the prop to ensure that the right usage of this API is promoted. There's an example of this for Popover's `anchor` prop that recommends particular usage - https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/popover/README.md#anchor-element--virtualelement--null.

> Also, because we have dropZoneRef?.current in the dependencies array of useRefEffect when the passed in ref changes, the cleanup appears to correctly fire, too

This is the part that often trips me up. With `dropZoneRef` being a ref type, it won't trigger a re-render when it changes. I've sometimes opted to store an element using `useState` for that reason (which is also what `Popover` recommends).

Reply:
> Sorry if I'm making this more complicated.

No apology necessary! I'm happy to go as slowly as needed with this PR — I always like to be a bit more cautious when making changes to lower-level packages such as `compose`, and it's important we get it right and don't introduce any issues 👍

> I don't know if there are tests that can be added to assert these different cases work as expected.

Good example use cases, it's worth digging in, I'll do a little more exploration to see if it's possible to come up with additional test cases. I think expanding the test cases there is probably a good path toward building confidence around the API change.

> It also might be worth expanding upon the docs for the prop to ensure that the right usage of this API is promoted. There's an example of this for Popover's anchor prop that recommends particular usage

Good idea, for some reason `useDropzone` isn't documented in the `compose` package's readme, but I'll have a go at writing an entry for it.

> This is the part that often trips me up. With dropZoneRef being a ref type, it won't trigger a re-render when it changes. I've sometimes opted to store an element using useState for that reason (which is also what Popover recommends).

Ah, good point, because the ref reference itself isn't changing, only `current` 🤔. In practice, I haven't yet run into any issues in this PR so far. In terms of using `useState`, do you mean we'd do that internally within `useDropZone`, or are you imagining that the consumer would use `useState`, and we'd pass the element directly to `useDropZone` instead of a `ref`?

Reply:
> Edit: I think you probably mean the latter — and I can see that the url popover for example, uses the state approach here

Yep, exactly that 👍

Reply:
Update: I've done the following:

* Swapped out `dropZoneRef` for a `dropZoneElement` and updated usage to use state for storing the element before passing it along to the `ListView` (and then on to `useDropZone`)
* Added a README file for `useDropZone` — because the hook is still currently exported as experimental, I couldn't add directly to the `compose` package readme. I see a couple of other experimental hooks also include their own README files in the `compose` package, so this seemed like a decent place to store the documentation for now, while the hook is still "experimental"

I haven't looked into exploring adding other tests just yet, but will dig in tomorrow.

Thanks again for the feedback!

Reply:
> because the hook is still currently exported as experimental, I couldn't add directly to the compose package readme.

Thanks, didn't realise it was experimental. It must have been that way for ages!

My hunch is that using state to store the element will probably solve the issues, hopefully that's the case.